### ISSUE TITLE
Infinite redirect loop in Rails 5.2

### ISSUE THREAD

**Issue by jacobwalkr** [nbudin/devise_cas_authenticatable#142]
Hi there,

I'm trying to move a Rails 5.2 (Devise 4.4.1) app from `devise_ldap_authenticatable` to `devise_cas_authenticatable` and am getting into a redirect loop once I have logged into the CAS server (`rubycas-server`). The whole setup is running in HTTP-only mode in Docker containers (with Docker Compose). It seems that an app session is never set, but I can't figure out why.

Full details are on [this StackOverflow post I wrote earlier](https://stackoverflow.com/questions/49110126/infinite-redirects-with-devise-and-devise-cas-authenticatable), but I wondered if someone here might have an idea. Even just some pointers for debugging would be very helpful - we've tried calling `byebug` before `authenticate_user!` but don't seem to have found anything useful.

Thanks!

**Comment by nbudin** [comment]
Based on the logs on StackOverflow, it looks like your CAS server is redirecting back to http://localhost:3000 without actually authenticating the user.  This often happens when the CAS server requires that each service must be explicitly registered, but this service hasn't been.  Which CAS server are you using?

**Comment by nbudin** [comment]
Never mind, I see that you are using rubycas-server (my reading comprehension is a little low today, sorry).  I haven't used rubycas-server myself, but from their wiki I don't see a feature that requires that each client be explicitly registered.

Nonetheless, the CAS server log you pasted in StackOverflow seems to indicate that it's just redirecting to the root without authenticating.  I'm not sure why this would be, but it's most likely an issue in the CAS server or how it's configured.

**Comment by jacobwalkr** [comment]
Hey, thanks for the response! It seems that CAS is authenticating properly through LDAP - I tried logging in by navigating directly to the CAS app:

![image](https://user-images.githubusercontent.com/2103207/37025844-adb10a26-2124-11e8-997c-0183be23b4d2.png)

It also seems to set a ticket _and_ verify it when asked. Which bit of the log were you looking at? Is the server issuing an invalid ticket?

**Comment by jacobwalkr** [comment]
To verify that the problem wasn't somewhere else in our setup, I created a fresh app with a fresh Docker setup, and it _worked_!

Notice that in the first log on my SO post, ActiveRecord selects the user first by username, then by email, which is null (because CAS doesn't send it).

If we comment out the email validation in the User model, it works as expected. The email field is to be fetched directly from LDAP in a `before_create` hook, to mimic `devise_ldap_authenticatable`'s `ldap_before_save` hook.

Could it be that Devise or `devise_cas_authenticatable` is doing some magic with model validations?

**Comment by nbudin** [comment]
Yeah, I think I may have been misreading your logs in SO.  Sorry about that!

I do think it's likely that somewhere in the chain, something is saving the user (and it's failing validations as a result).  My bet would be it's this line: https://github.com/nbudin/devise_cas_authenticatable/blob/master/lib/devise_cas_authenticatable/model.rb#L40

One thing you can do is put an `attr_accessor` on your user class to disable the email validation if present (something like `attr_accessor :skip_email_validation`).  You could then override `self.find_or_build_resource_from_conditions` to call super, set that attribute to true on the user, and return it.  That would let it pass validations temporarily, but keep the validation around for all subsequent calls.

**Comment by jacobwalkr** [comment]
Thanks! My understanding was that `before_create` came before validations - I was wrong.

I've fixed it by grabbing attributes before validations:

    before_validation :fetch_user_attributes, on: :create

Is this worth documenting in `devise_cas_authenticatable`? I don't know if anyone else is likely to come across this, but it has been a _right pain_ to figure out.

**Comment by jacobwalkr** [comment]
Is there a specific need to use the non-bang version of `resource.save`? Using `resource.save!` instead would raise an exception when validations fail, which would make the actual problem more apparent.

**Comment by jacobwalkr** [comment]
Sure, I'll think of something for the wiki! Thanks for your help :)
